### PR TITLE
use the `caml_example` environment in the Extensions manual

### PR DESCRIPTION
**Pull request by gasche**
This PR, which continues work started in #939 as tracked in [MPR#7551](https://caml.inria.fr/mantis/view.php?id=7551), has one well-defined commit that I believe is correct, but I marked it work-in-progress as:

- the whole section of the manual is not covered
- the parts that I did uncovered some minor details with the `{caml_example}{verbatim}` environment (from #1194) that we may want to discuss and potentially fix before going further

To that end I uploaded a [temporary rendering of the PR manual](http://www.ccs.neu.edu/home/gasche/tmp/htmlman/extn.html).

If you look at the [Recursive modules section](http://www.ccs.neu.edu/home/gasche/tmp/htmlman/extn.html#sec221) for example, you may notice that

1. The `#` at the beginning of the code example does not look nice (we are supposed not to be in the toplevel)
2. The spacing at the end of the code is too big. See also the examples in [private row types](http://www.ccs.neu.edu/home/gasche/tmp/htmlman/extn.html#sec225).
3. It would be sort of nice to have a way to show *signature item* examples that are syntax- and type-checked, but there is no urgency.

I think that it would be useful to think (cc @Octachron !) about fixing (1) and (2) before considering merging this or going further with the `caml_example`-ization of this section.

### CODE REVIEW

**Comment by Octachron** [comment]
For the first point, on the html side the `#` is implemented with a css rule. I would say the best way to remove it would be to add a `verbatim` (resp. `toplevel`) class to the `caml-example` `div`. This would allow to customize the style of both variant independently. On the latex side, removing the line of `#` on the first column might require some macro trickery. Note that the rendering of the output might benefit from some tuning too.

On the second point, the generated html code looks fine, this is probably a matter of tuning the margin/padding in the manual css style.

For 3, I think that wrapping the example in a module type definition before sending it to the toplevel might do the trick.

File: manual/manual/refman/exten.etex
Comment (Octachron) [inline manual/manual/refman/exten.etex]: May I suggest to add a star here `\end{caml_example*}` to not betray the illusion that caml_example is a latex environment? (pure nitpicking, I wholeheartedly admit)

File: manual/manual/refman/exten.etex
Comment (Octachron) [inline manual/manual/refman/exten.etex]: Another lost star above.

Comment (gasche) [reply manual/manual/refman/exten.etex]: (Thanks! Fixed both.)

**Comment by gasche** [comment]
I dislike LaTeX programming, so I don't plan to work on it. On the other hand I'll try to propose a patch for the html/css part along the lines proposed. (I'm interested in feedback on what future we want for this PR. Should we wait for the display issues to be fixed before considering a merge? Should we complete the porting of the chapter? (I would at least support the second option.))

**Comment by Octachron** [comment]
Completing the porting of the current section seems like a good idea: it increases the number of examples available for tuning the rendering afterwards. Moreover, since the fix for the rendering should be mostly independent from the manual changes, I don't see a lot of negatives with merging this PR before the rendering the fix. ((In term of negatives, I imagine that the rendering changes might require some micro-tuning of the manual examples)).

**Comment by xavierleroy** [comment]
Three months later, where are we on this GPR?

**Comment by gasche** [comment]
I haven't worked on it since, and I would rather not merge the change before the rendering gets improved. On the other hand, it's a low-conflict PR so it can stay open for longer. Thanks for the WIP tag, it's appropriate.

**Comment by Octachron** [comment]
@gasche , is there any obstacle (except the current conflict) remaining before merging?

**Comment by gasche** [comment]
I don't remember. I will fix the conflicts and rebase, and then maybe you could think about it, and if applicable accept and merge?

**Comment by gasche** [comment]
I'm looking at it now and it turns out that #1702 would be useful to go further on this PR.

**Comment by Octachron** [comment]
I have rebased #1702 (and moved the change entry to trunk).

**Comment by gasche** [comment]
I have rebased this PR, and added quite a few uses of caml_example. The remaining `verbatim` environments in exten.etex are for the cases where I don't know how to use caml_example instead, typically code that uses `...` for ellipsis -- and thus wouldn't compile.

**Comment by gasche** [comment]
Re dots.: a lot of the manual examples look like

```
let f (type a) (x : a -> a) = ...
```

@Octachron, as the manual/tools wizard, do you know if it would be possible to render this out of something like

```
let f (type a) (x : a -> a) = (assert false)[@ellipsis]
```
?

**Comment by Octachron** [comment]
It sounds possible. However, in order to have a robust behavior with attributes, the best choice is probably to parse each phrase and extract the location of each ellipsis from the parse tree.

**Comment by gasche** [comment]
This is what is done for expect-tests in testsuite/tools/expect_test.ml.

While we are at it, another random feature request inspired by this PR: if I pass `[error]` or `[warning=...]`, could I get the error or warning shown even in the `{caml_example*}` case? On my machine what I observe in the warning case is that the warning position is highlighted (a very nice feature in general), but the warning message is not shown.

Edit (August 2018): the question above has been fixed by #1863.

**Comment by gasche** [comment]
I just rebased this against #1765.

File: manual/manual/refman/exten.etex
Comment (gasche) [inline manual/manual/refman/exten.etex]: @Octachron I think that you may find this block of code interesting.

The first ellipsis is used in place of a separate `{caml_eval}` block as I had in my previous version (before the last rebase), and it suggests that maybe longer-term we could think of using `[@hide]` attributes that are completely removed instead of a separate environment.

The second ellipsis has this form, instead of a signature-item attribute, because of the issue I reported in https://github.com/ocaml/ocaml/pull/1765/files#r186431706. I am trying to come with a fix for the issue, but maybe we could consider trying to get this PR merged before that -- it's been long enough in the baking.

Comment (Octachron) [reply manual/manual/refman/exten.etex]: I agree that it would be nice to merge soon. I will do a full review by tomorrow.

> that maybe longer-term we could think of using [@hide] attributes that are completely removed instead of a separate environment.

This is definitively doable. At the same time, I fear that mixing hidden and non-hidden code fragment too freely might be confusing for the reader.

Comment (gasche) [reply manual/manual/refman/exten.etex]: Thinking more about it, I think we should wait until we agree on what to do with #1765 before merging.

**Comment by gasche** [comment]
I just rebased this PR on top of #1903, and the ellipsis machinery now works correctly in all examples where it is used.

@Octachron note that I added a patch, currently 48ea928cf451229b3439f1afa3e20f06287a243a, to simplify the location-handling in the ellipsis code, now that attributes carry their whole location.

File: manual/manual/refman/exten.etex
Comment (Octachron) [inline manual/manual/refman/exten.etex:2051]: Was this change by choice or due to the lack of attribute on case? If it is the second, using 
`| Other[@ellipsis.start] -> Other[@ellipsis stop]` would work after a small fix that I can send as a sub-PR.

File: manual/manual/refman/exten.etex
Comment (Octachron) [inline manual/manual/refman/exten.etex]: I would propose to have a separate ellipsis `[@@@ellipsis]` to have
```OCaml
module type DEVICe = sig
val draw:picture -> unit
...
end
```
rather than just `sig ... end` since it gives a clearer purpose to `picture`.

File: tools/Makefile
Comment (Octachron) [inline tools/Makefile:412]: I am curious, what is the story behind this `-no-alias-deps`?

**Review by Octachron — APPROVED** [review]
I have three small remarks but overall the examples look nice.

File: tools/Makefile
Comment (gasche) [reply tools/Makefile:412]: If you remove it, the build fails because `Parsetree` (to which an alias is introduced in #1903) has no `.cmo` to link against. This sounds strange, but `module M = Foo` is broken on all OCaml versions if `Foo` is mli-only and you don't use no-alias-deps.

File: manual/manual/refman/exten.etex
Comment (gasche) [reply manual/manual/refman/exten.etex:2051]: In that case I think that showing the full code is just as well. In your sub-PR, can you include a test somewhere? I should also add a regression test for `val picture : picture -> unit[@@ellipsis]` which I had a lot of trouble not to get rendered as `val picture : picture -> ...`.

**Comment by gasche** [comment]
I rebased on top of trunk, implementing @Octachron's recommendation (to show `val draw` in the `DEVICE` signature).

File: manual/manual/refman/exten.etex
Comment (Octachron) [reply manual/manual/refman/exten.etex:2051]: I think I will make a test PR for ellipsis attributes in general.

**Pull request merged**
